### ISSUE TITLE
Choice-free formatting via `ormolu` or `fourmolu` with default settings

### ISSUE THREAD

**Issue by GambolingPangolin** [jprupp/haskoin-core#399]
Is there any interest in formatting the code with `fourmolu` (or `ormolu`) to get a consistent formatting style?  IME this reduces developer effort, and compared to `stylish-haskell` reduces diff sizes.

**Comment by wraithm** [comment]
@jprupp, we've had great success with fourmolu and hls in our code base. ormolu would work great too!

**Comment by jprupp** [comment]
I'm happy with it, as long as I don't have to do editor hacking. I use Spacemacs and it doesn't seem to have direct support. Do I need editor support?

**Comment by wraithm** [comment]
Do you use lsp with spacemacs, as in haskell-language-server? Ormolu and fourmolu integrate with hls.

There is also ormolu.el (which works well with both ormolu and fourmolu) for emacs that's super straightforward to add. It's basically the same ergonomics as stylish-haskell's integration with haskell-mode.

You don't need editor support, but it's nice to have. Most IDEs/emacs/vim are covered with haskell-language-server. We could also add a script or something that calls the formatter, such that you don't have to care about editor setup.

What do you think?

**Comment by jprupp** [comment]
At the moment this seems like yak shaving. Let's revisit in the future as it matures and becomes better integrated with editors. By that I mean that there's a simple way to enable it in all editors, and it comes bundled with distributions or somehow it gets installed transparently without user effort.

An example of what it should be like: using `gofmt` when writing Go code. Once it is there, I'll have no problem using it.

**Comment by wraithm** [comment]
We have a formatter rn, it's stylish-haskell. Stylish-haskell isn't just automatically installed either. It's just that stylish-haskell is quickly becoming non-standard in the Haskell world.

Ormolu is super standard at this point. Like, rn, to work on Haskoin, I have to disable hls's standard formatting stuff and enable stylish-haskell via haskell-mode. All you have to do is `stack install ormolu` and then use haskell-language-server with that setting. I believe we can even put that setting in a .dir-locals.el, such that you don't even have to configure that setting at all if you're using emacs+hls.

Like, it basically is already at that point of being near automatic.

**Comment by jprupp** [comment]
There is support for Haskell Language Server in Spacemacs. I'll set it up. If it is straightforward and reasonably stable I'll use it. I'm having flashbacks to random issues with hindent that made it irritating to use. Let's see how this goes.

**Comment by wraithm** [comment]
You might be surprised! I remember those bad old days too. Ormolu/fourmolu
are waaaay better. It's basically at the gofmt level of ergonomics now.

**Comment by jprupp** [comment]
This is what I've done so far:

1. Read the disclaimer from the Spacemacs documentation for the develop version of the Haskell layer:

> lsp (hie or hls) is a more ambitious, heavy-weight, cutting-edge backend that is however still somewhat rough on the edges and requires some additional setup.

2. I don't like additional setup, but let's see what it entails. Another disclaimer:

>  hls is meant to supersede hie and is therefore the cutting-edge and most ambitious implementation of Haskell language backend. However, it is still in early stages of development.

3. Following the instructions, it tells me to go to the Haskell Language Server documentation:

> There are pre-built binaries available from the releases page for Linux, Windows and macOS. To install, download the haskell-language-server-wrapper executable for your platform as well as any haskell-language-server executables for the GHC versions you plan on working with, and either put them on your PATH or point your client to them.

Downloading a bunch of executable files (I have multiple Haskell projects on different versions of GHC), plus a wrapper and putting it in path doesn't sound fun. Let's see if there's an alternative:

> Binary packages for Fedora are available from this Copr repo, built against the official Fedora ghc package.

Hmm, apparently that will only work with GHC 8.8.4, as I'm running Fedora 34, so not good enough. Let's see if there's something else that I can do…

> Direct installation from Hackage, while possible via cabal install haskell-language-server, is not recommended for most people. Said command builds the haskell-language-server binary and installs it in the default Cabal binaries folder, but the binary will only work with projects that use the same GHC version that built it.

Nope, I'm definitely "most people" here. There are some options like building from source, which I won't get into.

I'm a very busy person. I need my tools to be easy to setup and use. It seems that this hls is heading in that direction, and I'm happy to revisit it once installation and configuration of the whole thing has been simplified to the point where I only have to flip a switch in the Haskell Layer, as I expect to be the case in a year or two. For now, I'll refrain from using it. There's still too much yak shaving.

**Comment by wraithm** [comment]
I'd really recommend using ghcup instead of what they say there. That tool
can set you up with hls. You don't have to worry about all of the
complexity that that documentation implies:

https://haskell.org/ghcup

If you install hls via their little "tui", you'll be good to go after
enabling hls in the Haskell layer of spacemacs. It's very easy now-a-days.

Also, if you're really busy, I'd be more than happy to take on more
responsibility with maintaining haskoin. If you give me hackage access, I
can make releases for you and such.

**Comment by jprupp** [comment]
I have been referring to yak shaving so far, and it is exactly what this is. I want a formatter, but end up with a new tool, ghcup, which so far has not been necessary for me to write Haskell code, another tool, hls, that is still in early development, and a strong suspicion that Spacemacs will start becoming wobbly after I activate all this.

I want a formatter, but I expect not to have to do anything fancy to set it up, that Spacemacs will automatically download and install the external tools it needs to get this going in a project-specific directory using Stack, as it does today with hindent if I enable it. That's the level of usability I expect. I have done enough editor and IDE warfare to know I rather avoid it. A formatter is nice to have, but not worth the hours spent setting it up and fixing it when it breaks, which I strongly suspect will happen on Linux distribution or Spacemacs upgrades.

Send me an email to get you setup in Hackage, and send you the release script that generates and uploads Haddock documentation along with the package.

**Comment by GambolingPangolin** [comment]
IMO making developers responsible for formatting is a drag on development that we should avoid.  `ormolu` completely automates formatting and leads to small diffs which is good for quality in an open project like this (it makes reviewing easier).  Moreover, I have some experience with how making this change significantly improved the development experience in a large industrial Haskell codebase.

The setup cost is very small.  There is no need to install `ghcup` or anything.  

1. Install the tool using whatever you usually use to install Haskell software.  (For me, `cabal install ormolu`.)
2. Configure your editor.  Emacs users can simply add `(use-package! ormolu)` or whatever and use a `.dir-locals.el` to turn on `ormolu-mode` in the project.

Haskell is not like go.  The culture and developer ecosystem is different.  Maybe there will be consensus on a formatter and it will wind up bundled with `GHC`, but probably not.  I feel it is a mistake to base this kind of decision on whether it is easy in Spacemacs, since that gives a lot of weight to a set of somewhat arbitrary things that Spacemacs users decided to automate.

**Comment by jprupp** [comment]
I managed to get `haskell-language-server` working on my system using its Fedora repository and direct support from Spacemacs. That's a level of complexity that I'm willing to tolerate, even if it means that I cannot bump my project's LTS Haskell version to use a newer compiler than what my distribution already provides, as it would break the IDE. So far, so good.

This formatter better be the cat's whiskers.

On Spacemacs, the only option available to me allows me to format the entire buffer (file) using the `lsp-format-buffer` "command". I believe that this is using `ormolu`.

I'd like to some guidance to validate that I'm on the right track. Here's a snippet of some formatted code:

```haskell
setMetrics ::
  MonadUnliftIO m =>
  (WebMetrics -> StatDist) ->
  WebT m ()
setMetrics df =
  asks webMetrics >>= mapM_ go
  where
    go m = do
      req <- S.request
      let t = fromMaybe e $ V.lookup (statKey m) (vault req)
      atomically $ writeTVar t (Just df)
    e = error "the ways of the warrior are yet to be mastered"
```

Does that look like `ormolu` format to you?

**Comment by GambolingPangolin** [comment]
Yes, your snippet looks like it's an output of `ormolu`.  HLS and `ormolu` are related but if you are reluctant to tie yourself to a supported GHC version, you can install `ormolu` as a standalone executable and use `ormolu-mode`.

**Comment by GambolingPangolin** [comment]
HLS is def worth exploring.  I find it very useful!  For example, it implements go to definition (in your project) and some nice code completions like branching on a sum type or giving you a record literal definition.

**Comment by jprupp** [comment]
Yes, I also found an `ormolu` package in Fedora repos. Now I'm confident that I can help someone else set this up on his computer to contribute code.

Now I'll reformat the entire project and push it on a single large refactoring commit.

Tell me if you have anything that is pending to be merged before I do this.

-------- Original Message --------

**Comment by GambolingPangolin** [comment]
https://github.com/haskoin/haskoin-core/pull/402 is just waiting on a response from @wraithm or @jprupp about the signer API, but I'm happy to rebase if you don't wanna wait for it.

**Comment by wraithm** [comment]
> https://github.com/haskoin/haskoin-core/pull/402 is just waiting on a response from @wraithm or @jprupp about the signer API, but I'm happy to rebase if you don't wanna wait for it.

Merged

**Comment by wraithm** [comment]
Yeah, I'm sorry that I didn't fully explain what's going on @jprupp. hls has an ormolu plugin, but ormolu can absolutely stand on its own. There's the ormolu.el emacs package, but you don't need that if you use hls, which is what everybody is using these days. hls is also editor agnostic! hls and its ormolu formatter works just as well in emacs, vim, vscode, etc etc!

It really is a huge step in the direction of that frictionless set-up that you're looking for.
